### DISCUSSION TITLE
Image Picker State

### DISCUSSION THREAD

**Discussion by caitlynbakery** [caitlynbakery/12_take_save_display#1] [General]
In Project 11, I didn't use a Bloc structure for image_picker. However, in Project 12, I use a Bloc structure to handle the multiple states. The main problem I'm encountering is what code should be handled inside of the Bloc for picking an image within the image gallery. As the image file is updated when the user clicks on a new image, ideally the Bloc would manage the image. Unfortunately, I ran into some issues with trying to manage the image file with the Bloc. 

https://github.com/caitlynbakery/12_take_save_display/blob/cf20c833a8d5bb3343f9c05737287cf2608f5836/lib/blocs/theta/theta_state.dart#L10

@codetricity

**Comment by codetricity** [comment]
https://github.com/hprity60/multi-image-picker-bloc-8
https://youtu.be/qLXSZOmTlZ4

she seems to have more videos here and you can see what she looks like too.

https://www.youtube.com/c/LearnFlutterWithSmrity

She uses a question mark after the list type for null safety

---

```dart
part of 'product_bloc.dart';

@immutable
class ProductState {
  final List<XFile>? images;

  ProductState({this.images});

  ProductState copywith({List<XFile>? images}) =>
      ProductState(images: images ?? this.images);
}
```

## pass images to event

```dart
        child: InkWell(
          onTap: () async {
            final ImagePicker _picker = ImagePicker();
            final List<XFile>? images = await _picker.pickMultiImage();
            if (images != null) {
              productBloc.add(onSelectMultipleImageEvent(images));
            }
          },
```

## product_event.dart

```dart
part of 'product_bloc.dart';

@immutable
abstract class ProductEvent {}

class onSelectMultipleImageEvent extends ProductEvent {
  final List<XFile> images;

  onSelectMultipleImageEvent(this.images);
}

class onUnSelectMultipleImageEvent extends ProductEvent {}
```

## product_bloc.dart

```dart
  Future<void> _onSelectMultipleImages(
      onSelectMultipleImageEvent event, Emitter<ProductState> emit) async {
    emit(state.copywith(images: event.images));
  }
```

**Reply by caitlynbakery** [reply]
I got the image_picker working using the example you provided above. Currently, the app can pick the image from the image gallery using Bloc. However, I still need to add view image feature.

Here is how the app looks now:

![appdemo](https://user-images.githubusercontent.com/22283010/181786957-d32ae42e-43b1-4225-b110-bd4568fb99cc.gif)
 
I am on this step in the diagram:

<img width="521" alt="Screen Shot 2022-07-29 at 7 56 04 AM" src="https://user-images.githubusercontent.com/22283010/181787228-d68fa1c4-6781-4a95-a4b9-3828ae926ba1.png">

<img width="933" alt="Screen Shot 2022-07-29 at 7 57 38 AM" src="https://user-images.githubusercontent.com/22283010/181787530-30b78488-7a7a-4f32-9b79-413fdf5a8325.png">

**Reply by codetricity** [reply]
> I got the image_picker working using the example you provided above. Currently, the app can pick the image from the image gallery using Bloc. However, I still need to add view image feature.

If you can continue to use the BLoC architecture, please use it.  Although it may be more confusing initially, I hope that we will gain more experience.  I hope that using BLoC will become easier.

It looks like you're using the technique of putting a question mark at the end of the data type.

```
 final XFile? images;
```

https://github.com/caitlynbakery/12_take_save_display/blob/b24729cc35a9761fd451b229d20ba072fe0af5d9/lib/blocs/theta/theta_state.dart#L10

I notice that your copyWith method is different from Learn Flutter With Smrity example.

Current `copyWith` only has `message`, not the other variables

```dart
  ThetaState copyWith({
    String? message,
  }) {
    return ThetaState(
      message: message ?? this.message,
    );
  }
```

This might be another way to structure `theta_state.dart` which might force us to research what the `copyWith` actually does.

```dart
part of 'theta_bloc.dart';

class ThetaState extends Equatable {
  final String message;
  final String fileUrl;
  final String cameraState;
  final String id;
  final bool finishedSaving;
  final XFile? imageFile;

  ThetaState(
      {required this.message,
      this.fileUrl = "",
      this.cameraState = "initial",
      this.id = "",
      this.finishedSaving = false,
      this.imageFile});

  factory ThetaState.initial() => ThetaState(message: "Response from Camera");

  @override
  List<Object> get props => [message, id, cameraState, finishedSaving];

  @override
  bool get stringify => true;

  ThetaState copyWith({
    String? message,
    String? fileUrl,
    String? cameraState,
    String? id,
    bool? finishedSaving,
    XFile? imageFile,
  }) {
    return ThetaState(
      message: message ?? this.message,
      fileUrl: fileUrl ?? this.fileUrl,
      cameraState: cameraState ?? this.cameraState,
      id: id ?? this.id,
      finishedSaving: finishedSaving ?? this.finishedSaving,
      imageFile: imageFile ?? this.imageFile,
    );
  }
}
```

## what is emit `state.copyWith()`?

The hprity60 (learn with smrty) example uses this technique

```dart
class ProductBloc extends Bloc<ProductEvent, ProductState> {
  ProductBloc() : super(ProductState()) {
    on<onSelectMultipleImageEvent>((_onSelectMultipleImages));
    on<onUnSelectMultipleImageEvent>((_onUnSelectMultipleImages));
  }
  Future<void> _onSelectMultipleImages(
      onSelectMultipleImageEvent event, Emitter<ProductState> emit) async {
    emit(state.copywith(images: event.images));
  }

  Future<void> _onUnSelectMultipleImages(
      onUnSelectMultipleImageEvent event, Emitter<ProductState> emit) async {
    emit(state.copywith(images: []));
  }
}
```

If possible, explain in this discussion what the syntax `state.copyWith` does in the hprity60 example if it beneficial to refactor your code to use a similar strategy.

---

https://github.com/caitlynbakery/12_take_save_display/blob/b24729cc35a9761fd451b229d20ba072fe0af5d9/lib/blocs/theta/theta_bloc.dart#L48

If you put the variables in copyWith, I think you can simplify the state in the bloc.

```dart
        emit(ThetaState(
            message: response.body, id: state.id, cameraState: cameraState));
      }
```

can  be

```dart
emit(state.copyWith(message: response.body);
```

**Reply by caitlynbakery** [reply]
I'll experiment and change the emit statements to copyWith in the Bloc file.

**Reply by jcasman** [reply]
Caitlyn, nice progress! I find the graphic very useful in seeing what you're programming. How long did it take to make the graphic in Figma? Did you revise it much?

**Reply by codetricity** [reply]
@caitlynbakery , after discussion, I understand that you didn't attempt to change the stream event to  `emit(state.copyWith(message: response.body, ...`

I still think it is worthwhile to give it a try next time and decide if it is a better syntax.  The reason I suggest you experiment with it is that the author of the tutorial, smrty used the technique of `state.copyWith...`

**Reply by caitlynbakery** [reply]
@jcasman It took me around 30 minutes to make the diagram in Figma and I didn't revise it afterwards. It's still applicable to my project for now. At the moment, I don't need to change it.

**Reply by caitlynbakery** [reply]
@codetricity I changed 3/5 of the Blocs to use `state.copyWith`. The advantage seems to be that it is more clear which state changes and there is less code.

**Reply by codetricity** [reply]
nice work figuring out how to use copyWith.  Suggest experimenting with both `emit(state.copyWith()` and emit(ThetaState())` going forward.  Choose the technique to meet your requirements.
